### PR TITLE
Exchange client cleanup

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/execution/SharedBufferMemoryManager.java [hunk 1]
```diff
@@ -30,12 +40,36 @@ public SharedBufferMemoryManager(long maxBufferedBytes, SystemMemoryUsageListene
...
+
+    public synchronized ListenableFuture<?> getNotFullFuture()
+    {
+        if (isFull() && notFull.isDone()) {
+            notFull = SettableFuture.create();
+        }
+        return notFull;
+    }
+
+    public void setNoBlockOnFull()
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/execution/SharedBufferMemoryManager.java]: What's the purpose of this?

Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/execution/SharedBufferMemoryManager.java]: When the buffer is destroyed (say due to a limit), we need free the writers (and make sure no one else blocks).  This is the least error prone method I could think of to accomplish this.

File: presto-main/src/main/java/com/facebook/presto/operator/HttpPageBufferClient.java [hunk 2]
```diff
@@ -275,9 +275,9 @@ private synchronized void initiateRequest()
         lastUpdate = DateTime.now();
     }
 
-    private void sendGetResults()
+    private synchronized void sendGetResults()
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/HttpPageBufferClient.java]: Add a comment to the commit explaining why. This doesn't look like just fixing a warning.

File: presto-main/src/main/java/com/facebook/presto/execution/PageBufferInfo.java [hunk 3]
```diff
@@ -26,6 +26,7 @@
     private final long bufferedPages;
     private final long queuedPages;
     private final long bufferedBytes;
+    private final long rowsAdded;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/execution/PageBufferInfo.java]: The commit message doesn't seem to match what this is tracking (i.e., rows added to the buffer vs rows sent to the client)

File: presto-main/src/main/java/com/facebook/presto/execution/SharedBufferMemoryManager.java [hunk 4]
```diff
@@ -22,6 +28,10 @@
 {
     private final long maxBufferedBytes;
     private final AtomicLong bufferedBytes = new AtomicLong();
+    @GuardedBy("this")
+    private SettableFuture<?> notFull;
+
+    private final AtomicBoolean noBlockOnFull = new AtomicBoolean();
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/execution/SharedBufferMemoryManager.java]: The code may be easier to follow if you make this a positive: `blockOnFull`

**Comment by martint** [comment]
Looks good, but defer the first commit until after we complete this release.

**Pull request closed without merging**